---
title: "Proyecto final Genómica Humana: Análisis de Mutaciones mendelianas"
date: "7/06/2021"
author: "Alfredo Varela Vega, Elizabeth Márquez Gómez, Joshua Rodríguez Vázquez, Omar Daniel López Olmos"
output: 
  html_document:
    df_print: paged
    code_folding: show
    self_contained: yes
    theme: yeti
    highlight: haddock
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: "3"
---


```{r}
library(dplyr)
library(biomaRt)
library(ggplot2)
library(tidyverse)
library(rentrez)
library(scales)
library(ggrepel)
library(RColorBrewer)
```



### Leer el archivo de [OMIM](https://www.omim.org) que contiene los fenotipos e identificadores de ENSEMBL

```{r}
# Reading the file 
omim_genes<- read.table("data/genemap2.txt", sep = "\t", header= FALSE, fill =TRUE)
# Creating the header 
colnames(omim_genes) <- c("chromosome",	"genomic_position_start", "genomic_position_end", "cyto_location","computed_cyto_location", "mim_number","gene_symbols","gene_name","approved_symbol","gene_id","ensembl_gene_id","comments","phenotypes","mouse_geneSymbol/ID")

head(omim_genes)
```

### Filtar enfermedades mendelianas (autosómico dominante, autosómico recesivo, X-linked y Y-linked)

```{r}
mendelian_omim_genes<- omim_genes %>%
  # Looking for autosomal dominant, autosomal recessive, X-linked dominaant
   filter(grepl("Autosomal|X-linked|Y-linked", phenotypes, ignore.case = TRUE))

# Checking duplicated values from mim id
#head(unique(mendelian_omim_genes$mim_number[duplicated(mendelian_omim_genes$mim_number)]))
# Checking duplicated diseases
# head(unique(query_request_3$phenotype_description[duplicated(query_request_3$phenotype_description)]))
head(mendelian_omim_genes)
```

### Búsqueda en biomart de ENSEMBL por medio del MIM number para anotar la posición los loci en genoma de [homo sapiens](https://www.ensembl.org/Homo_sapiens/Info/Index) 

```{r}
# conncet to genes database
ensembl <- useEnsembl(biomart = "genes")


#	Check available data sets, hsapiens_gene_ensembl corresponde a Human genes (GRCh38.p13)
datasets <- listDatasets(ensembl)
head(datasets)

# accesing homo sapines genes
ensembl <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl)

ensembl
# filters to apply to my query 
#listFilters(ensembl)
# output attributes i want 
#listAttributes(ensembl)

list_Attributes<- as.data.frame(listAttributes(ensembl))

# getting biotypes for all our previous omim genes
query_request_3 <- getBM(attributes=c("ensembl_gene_id","chromosome_name","gene_biotype","mim_gene_accession","phenotype_description"),filters=c("mim_gene_accession"),values=mendelian_omim_genes$mim_number, mart= ensembl)
cat("Obtener los biotipos para los genes obtenidos de OMIM: \n")
head(query_request_3)
# Filtering only for correctly annotated chromosomes from 1 to 21, X and Y
query_request<- query_request_3 %>% 
  filter(chromosome_name %in% c(1:21)| chromosome_name == c("X")| chromosome_name == c("Y")) 
cat("Filtrado por cromosoma: \n")
head(query_request)
```

### Datos non coding 

Leyendo los datos de la [base de datos non-coding](https://github.com/Gardner-BinfLab/ncVarDB).
[ncVarDB](https://academic.oup.com/database/article/doi/10.1093/database/baaa105/6013764)

```{r}
non_cod_data<- read.table("data/nc_patho_variants/ncVar_pathogenic.tsv", sep ="\t", header = TRUE)
head(non_cod_data)
```

### Visualización de los resultados conjuntos de ncVarDB y OMIM

#### 1. Pie de proporciones de biotipos de mutación 

```{r}
# Creating frequency tables and joining theme by rows 
ensembl_freq<-as.data.frame(table(query_request$gene_biotype))
non_cod_freq<-as.data.frame(table(non_cod_data$mut_position))
biotype_counts<-rbind(ensembl_freq,non_cod_freq)

biotype_counts

# Biotype Pie chart 
#png("./results/mendelian_biotype_pie.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(biotype_counts, aes(x ="", y = Freq, fill = Var1)) + 
  geom_bar(stat = "identity") + 
  coord_polar("y", start = 0) + 
  theme(axis.text = element_blank(), 
        axis.ticks = element_blank(), 
        panel.grid = element_blank()) + 
  labs(title = "Proporciones de biotipos en mutaciones Mendelianas", x= "", y="") + 
  scale_fill_brewer(name = "Posición de la mutación", palette = "Paired")
#dev.off()
```


#### 2. Gráfica de barras de non-coding 

```{r}
# Using all biotypes but protein coding 
biotype_counts_no_prot<-data.frame(biotype= biotype_counts$Var1[biotype_counts$Var1 != "protein_coding"], frequency = biotype_counts$Freq[biotype_counts$Var1 != "protein_coding" ] )

biotype_counts_no_prot<- biotype_counts_no_prot %>%
  arrange(desc(frequency))
# Non coding bar plot 
#png("./results/no_prot_mendelian_biotype_bar_plot.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(as.data.frame(biotype_counts_no_prot), aes(x = reorder(biotype,frequency), y = frequency , fill= biotype)) +
  geom_bar(stat = "identity") + 
  coord_flip() + 
  labs(title = "Posiciones de las mutaciones en enfermedades mendelianas", x= "Posición", y = "Frecuencia") + 
  geom_label(label=biotype_counts_no_prot$frequency) + 
  scale_fill_brewer(palette = "Paired") +
  theme_classic() + 
  theme(legend.position = "none")
#dev.off()

```

#### 3. Mutaciones por cromosoma

No todas las mutaciones tenían correcta la anotación del cromosoma al que correspondían por lo tanto solo se graficaron los que contenían la anotación correcta. la suma total de observaciones es menor a la de las visualizaciones por localización de la mutación debido a lo anterior. 

```{r}
# Creating the chromosome counts data frame
non_cod_chr_count<-as.data.frame(table(non_cod_data$chr))
omim_chr_count<-as.data.frame(table(query_request$chromosome_name))
# Merging the omim and the ncVarDB data frames
chr_merged<-merge(non_cod_chr_count, omim_chr_count, by.x ="Var1", by.y = "Var1", all.x = TRUE, all.y = TRUE)
#Change column names
colnames(chr_merged)<- c("chromosome", "Count_ncVarDB","Count_OMIM")
# Changing Na's for 0
chr_merged[is.na(chr_merged)] <- 0

# Create total count of mutations
chr_merged<- chr_merged %>%
  mutate(total_mutations = Count_ncVarDB + Count_OMIM) %>%
   arrange(desc(total_mutations))

n_colors<- length(chr_merged$chromosome)
mycolors <- colorRampPalette(brewer.pal(12, "Paired"))


#png("./results/mutations_per_chromosome_bar_plot.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(chr_merged, aes(x =reorder(chromosome,total_mutations) , y = total_mutations, fill = chromosome))+
  geom_bar(stat="identity") + 
  labs(title="Frecuencia de mutaciones por cromosoma", x= "Cromosoma", y = "Frecuencia") + 
  geom_label(label=chr_merged$total_mutations) +
  theme_classic() + 
  theme(legend.position = "none") +
  coord_flip() + 
  scale_fill_manual(values= mycolors(n_colors))
#dev.off
```


#### 4. Por tipo de herencia y sitio codificante o no codificante

El tipo de herencia se extrajo en las mutaciones que lo contenían. Muchas de las mutaciones que caen en sitios no codificantes carecen de la anotación correcta del tipo de herencia. Por ello  solo se graficaron los que contenían la anotación correcta de ambas variables. La suma total de observaciones es menor a la de las visualizaciones por localización de la mutación debido a lo anterior. 

```{r}
# Autosomal dominant from ncVarDB
non_cod_auto_dom<- non_cod_data %>%
  filter(grepl("Autosomal dominant", phenotype, ignore.case = TRUE)) %>%
  dplyr::select(phenotype) %>%
  mutate(inheritance = c("autosomal dominant"), 
         type = c("non-coding"))
# Autosomal recessive from ncVarDB
non_cod_auto_rec<- non_cod_data %>%
  filter(grepl("Autosomal recessive", phenotype, ignore.case = TRUE)) %>%
  dplyr::select(phenotype) %>%
  mutate(inheritance = c("autosomal recessive"), 
         type = c("non-coding"))

# there are no x linked dominants in non_cod_data so only x-linked recessive form ncVarDB neither Y-linked
non_cod_x_rec<- non_cod_data %>%
  filter(grepl("X-linked recessive", phenotype, ignore.case = TRUE)) %>%
  dplyr::select(phenotype) %>%
  mutate(inheritance = c("x-linked recessive"), 
         type = c("non-coding"))

# omim autosomal dominant 
omim_auto_dom<- mendelian_omim_genes %>%
  filter(grepl("Autosomal dominant", phenotypes, ignore.case = TRUE)) %>%
  dplyr::select(phenotypes) %>%
  mutate(inheritance = c("autosomal dominant"), 
         type = c("coding"))
# omim autosomal recessive
omim_auto_rec<- mendelian_omim_genes %>%
  filter(grepl("Autosomal recessive", phenotypes, ignore.case = TRUE)) %>%
  dplyr::select(phenotypes) %>%
  mutate(inheritance = c("autosomal recessive"), 
         type = c("coding"))
# omim X-linked dominant 
omim_x_linked_dom<- mendelian_omim_genes %>%
  filter(grepl("X-linked dominant", phenotypes, ignore.case = TRUE)) %>%
  dplyr::select(phenotypes) %>%
  mutate(inheritance = c("x-linked dominant"), 
         type = c("coding"))
# omim X-linked recessive 
omim_x_linked_rec<- mendelian_omim_genes %>%
  filter(grepl("X-linked recessive", phenotypes, ignore.case = TRUE)) %>%
  dplyr::select(phenotypes) %>%
  mutate(inheritance = c("x-linked recessive"), 
         type = c("coding"))
# omim Y-linked because no distinction dominant or recessive 
omim_y_linked <- mendelian_omim_genes %>%
  filter(grepl("Y-linked", phenotypes, ignore.case = TRUE)) %>%
  dplyr::select(phenotypes) %>%
  mutate(inheritance = c("Y-linked recessive"), 
         type = c("coding"))

inheritance_df <- bind_rows(non_cod_auto_dom, non_cod_auto_rec, non_cod_x_rec, omim_auto_dom, omim_auto_rec, omim_x_linked_dom, omim_x_linked_rec, omim_y_linked)
```

```{r}
# Creating data frame of counts for inheritance
inheritance_counts<- as.data.frame(table(inheritance_df$inheritance))
counts <-as.data.frame(tapply(inheritance_df$inheritance, list(inheritance_df$inheritance, inheritance_df$type), table))

# Adding receptor names columns 
counts$inheritance<-rownames(counts)

# Change to long format 
counts_long <- reshape(counts, 
                       direction = "long", 
                       varying = list(names(counts)[1:2]),
                       v.names = "Counts", 
                       idvar = "inheritance", 
                       timevar = "Status",
                       times=c("coding","non-coding"))


# Replace NA with 0 
counts_long[is.na(counts_long)] <- 0
counts_long

#png("./results/inheritance_bar_plot.png", width = 4, height = 4, units = 'in', res = 300)
ggplot(counts_long, aes(x = inheritance , y = Counts, fill = Status))+
  geom_bar(stat="identity") + 
  labs(title="Tipo de herencia de las mutaciones", x= "Herencia", y = "Frecuencia") + 
  theme_classic() + 
  coord_flip() + 
  scale_fill_brewer(palette = "Paired")
#dev.off
```



